reconcile/status_page_components.py

```diff
@@ -0,0 +1,261 @@
+    """
+    app-interface schema dependencies/status-page-1
+    """
+
+    name: str
+    page_id: str = Field(..., alias="pageId")
+    api_url: HttpUrl = Field(..., alias="apiUrl")
+    credentials: VaultSecretRef
+    components: list[StatusComponent]
+    provider: str
+
+    def get_page_provider(self) -> StatusPageProvider:
+        if self.provider == "atlassian":
+            return AtlassianStatusPage(page_id=self.page_id,
+                                       api_url=self.api_url,
+                                       token=self.credentials.get("token"))
+        else:
+            raise ValueError(f"provider {self.provider} is not supported")
+
+    def reconcile(self, dry_run: bool, state: State):
+        name_to_id_state = state.get_all("")
+        page_provider = self.get_page_provider()
+
+        # restore component ids from state
+        for desired in self.components:
+            desired.component_id = name_to_id_state.get(desired.name)
+
+        # delete
+        id_to_name_state = {v: k for k, v in name_to_id_state.items()}
+        desired_component_names = [c.name for c in self.components]
+        for current_id in page_provider.component_ids():
+            # if the component is known to the state management and if it is
+            # not known to the desired state, it was once managed by this
+            # integration but was delete from app-interface -> delete from page
+            name_for_current_component = id_to_name_state.get(current_id)
+            if (name_for_current_component and
+               name_for_current_component not in desired_component_names):
+                LOG.info(f"delete component {name_for_current_component} "
+                         f"from page {self.name}")
+                page_provider.delete_component(dry_run, current_id)
+                state.rm(name_for_current_component)
+
+        # create and update
+        for desired in self.components:
+            component_id = page_provider.apply_component(dry_run, desired)
+            if component_id and desired.component_id != component_id:
+                self._bind_component(dry_run, desired, component_id, state)
+
+    def _bind_component(self, dry_run: bool,
+                        component: StatusComponent, component_id: str,
+                        state: State) -> None:
+        LOG.info(f"bind component {component.name} to ID {component_id} "
+                 f"on page {self.name}")
+        if not dry_run:
+            state.add(component.name, component_id, force=True)
+            component.component_id = component_id
+
+
+class AtlassianStatusPage(StatusPageProvider):
+
+    page_id: str
+    api_url: str
+    token: str
+
+    components_by_id: dict[str, AtlassianComponent] = {}
+    components_by_displayname: dict[str, AtlassianComponent] = {}
+    group_name_to_id: dict[str, str] = {}
+
+    def __init__(self, **data):
+        super().__init__(**data)
+        self._rebuild_state()
```

Comment:
Can we rebuild the state outside of `__init__`? I see this ends up calling `self._client()` which means the constructor will have side effects, which makes modification and testing harder. We document it [in the style guide](https://service.pages.redhat.com/dev-guidelines/docs/appsre/python-style-guide/#avoid-side-effects-in-constructors) and yes, I know, we have a lot of code breaking this convention but this is a good time to move to a better style.

Reply:
fixed